routing/control_tower.go

```diff
@@ -305,4 +352,18 @@ func (p *controlTower) notifySubscribers(paymentHash lntypes.Hash,
 		case <-subscriber.quit:
 		}
 	}
+
+	// Notify all subscribers that subscribed to all payments.
+	for key, subscriber := range p.subscribersAllPayments {
```

Comment:
We still read from a map that could be written to concurrently. I think we need to hold the lock, create a local copy, then release the lock again and loop over the copy instead. Then only lock again on delete (as it's currently implemented on L364 and following).

Reply:
We can see the problem [here](https://go.dev/play/p/eDysaEIMGGA)

Reply:
I do agree this is the case. But there's a note on the rpc call saying:

> Note that if payments are in-flight while starting a new subscription, the start of the payment stream could produce out-of-order and/or duplicate events. In order to get updates for every in-flight payment attempt make sure to subscribe to this method before initiating any payments.

That effect doesn't change whether we hold the lock and copy the map locally or not. It will still depend on timing whether you receive a duplicate or not. (Because the inflight HTLCs are fetched too when you subscribe to trackpayments)

So I'd prefer to not copy the thing over to save some cpu and time in a global lock.

Reply:
This could potentially end up in a panic by the go runtime (Concurrent Map Misuse)

```
fatal error: concurrent map iteration and map write
```

To make it even worst, this is a `throw` not a `panic` which means there is no way to recover. The only solution here is to actually copy the map and iterate or use a lock.

Reply:
OK! I'm convinced. Done.